### PR TITLE
UserHyph: Scrub and sort user dict

### PR DESCRIPTION
**Pull request by zwim**
This should fix https://github.com/koreader/koreader/issues/8843

The problem was, that some user dictionary entries were not utf8-normalized.

With this PR, old dicts get normalized on load and if there is a problem with loading the dicts are get sorted again.
This sorting might help, if someone generates his dicts per hand too.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8863)

### CODE REVIEW

**Comment by zwim** [comment]
@ichnilatis-gr If this PR is merged, you can use your `user-Greek.hyph` again.

**Review by Frenzie — APPROVED** [review]

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 1]
```diff
@@ -170,6 +187,54 @@ function ReaderUserHyph:updateDictionary(word, hyphenation)
     self:loadUserDictionary(true) -- dictionary has changed, force a reload here
 end
 
+-- This function should only be called, if a user edits the hyphenation file per hand
+-- and messes the order
+-- The dictionary file will not get reloaded here!
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:192]:
```suggestion
-- This is called when the file is badly sorted (which should only happen if a user has edited
-- the hyphenation file by hand and messed the order)
-- Dictionary reloading is to be done by the caller (is this worth being said ? with a "!" the caller being this file :)
```

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 2]
```diff
@@ -170,6 +187,54 @@ function ReaderUserHyph:updateDictionary(word, hyphenation)
...
+        return
+    end
+
+    local dict_entries = {}
+
+    local line = dict:read()
+    if line then
+        line = Utf8Proc.normalize_NFC(line)
+    end
+    --search entry
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:208]: we're not searching an entry (?)

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 3]
```diff
@@ -120,16 +126,26 @@ function ReaderUserHyph:updateDictionary(word, hyphenation)
...
+
     local word_lower = Utf8Proc.lowercase(word)
     local line
 
     local dict = io.open(dict_file, "r")
     if dict then
         line = dict:read()
+        if line then
+            line = Utf8Proc.normalize_NFC(line)
+        end
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:141]: Would `line = line and Utf8Proc.normalize_NFC(line)` just work (and keep all these single line) ?

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 4]
```diff
@@ -28,26 +28,32 @@ end
...
         local ret = cre.setUserHyphenationDict(name, reload)
         -- this should only happen, if a user edits a dictionary by hand or the user messed
         -- with the dictionary file by hand. -> Warning and disable.
         if ret == self.USER_DICT_ERROR_NOT_SORTED then
-            UIManager:show(InfoMessage:new{
-                text = T(_("The user dictionary\n%1\nis not alphabetically sorted.\n\nIt has been disabled."), name),
-            })
-            logger.warn("UserHyph: Dictionary " .. name .. " is not sorted alphabetically.")
-            G_reader_settings:makeFalse("hyph_user_dict")
+            if _second_try == true then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:39]: no need for `== true`

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 5]
```diff
@@ -28,26 +28,32 @@ end
 -- Load the user dictionary suitable for the actual language
 -- if reload==true, force a reload
 -- Unload is done automatically when a new dictionary is loaded.
-function ReaderUserHyph:loadDictionary(name, reload)
+-- _second_try is reserved internally if the user dictionary gets scrubbed/sorted
+function ReaderUserHyph:loadDictionary(name, reload, _second_try)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:32]: No need for this comment, and no need for the underscore. It's all internal :)
Just name it `no_scrubbing`, so it documents itself.

**Comment by poire-z** [comment]
> So, do I need to check the "greek_block" file you gave me?

Yes you do :)
This greek block is for crengine itself - if we manage to get all this correct, it's possible you won't need to use "user hyphenation" as much as currently, and/or it will work better.

**Comment by zwim** [comment]
Ready to merge?

**Comment by ichnilatis-gr** [comment]
I'm sorry, I haven't checked Greek block yet due to lack of time...

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 6]
```diff
@@ -120,16 +125,22 @@ function ReaderUserHyph:updateDictionary(word, hyphenation)
         return
     end
 
+    if word then
+        word = Utf8Proc.normalize_NFC(word)
+    end
+
     local word_lower = Utf8Proc.lowercase(word)
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:132]: Won't the 2nd one failed if no word ? If not, is it different for the firsts one ?
May be shorten the first one like you did for          `line = line and Utf8Proc.normalize_NFC(line)`

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 7]
```diff
@@ -28,26 +28,31 @@ end
...
         -- with the dictionary file by hand. -> Warning and disable.
         if ret == self.USER_DICT_ERROR_NOT_SORTED then
-            UIManager:show(InfoMessage:new{
-                text = T(_("The user dictionary\n%1\nis not alphabetically sorted.\n\nIt has been disabled."), name),
-            })
-            logger.warn("UserHyph: Dictionary " .. name .. " is not sorted alphabetically.")
-            G_reader_settings:makeFalse("hyph_user_dict")
+            if no_scrubbing then
+                UIManager:show(InfoMessage:new{
+                    text = T(_("The user dictionary\n%1\nis not alphabetically sorted.\n\nIt will be disabled now."), name),
```
Comment (poire-z) [inline frontend/apps/reader/modules/readeruserhyph.lua:40]: Is it possible that after "scrubbing", you'd still get the user dict not sorted ?

**Comment by poire-z** [comment]
> I'm sorry, I haven't checked Greek block yet due to lack of time... I'll try over the weekend.

That's unrelated to this PR work, so this shouldn't prevent this PR from being merged (and solved your sorting issue).
No hurry with the greek block inspection & fix, that's for integration into crengine, and I'm a bit too busy to work on this soon.

File: frontend/apps/reader/modules/readeruserhyph.lua (hunk 7 above)
Comment (zwim) [reply frontend/apps/reader/modules/readeruserhyph.lua:40]: I hope that we never see that message. If we see that message, we have a bug: Either in `scrubDictionary()` or in crengine's `UserHyphDict::addEntry`.

File: frontend/apps/reader/modules/readeruserhyph.lua (hunk 6 above)
Comment (zwim) [reply frontend/apps/reader/modules/readeruserhyph.lua:132]: The check is not necessary.

**Review by poire-z — APPROVED** [review]

**Comment by zwim** [comment]
Please don't merge, there I have possibly found a bug.

**Comment by zwim** [comment]
Don't merge as there is a problem with hyphenation now. Has something to do with utf8_normalize.

**Comment by zwim** [comment]
OK, I got it now.
In Lua-land our utf8proc functions use utf8proc_NFC and for lowercasing utf8proc_NFKC_casefold.

-> 1st question not directly related to the problem at hand: Why don't we always use NFKC? On the first view this would make sense. Could someone of the knowing help me to get the difference between NFC and NFKC? 

In creninge-land there is nothing like normalization :( 
So we have a problem, when our dictionary entries are normalized, but the strings to hyphenate are not.

-> Solution: Crengine has to normalize when doing any of its `Utf8ToUnicoe` methods (but thats simple, as all end in `DecodeUtf8`)

A q&d PR on that will follow in crengine. With that PR it is possible to hyphenate words from hius7's `user-Greek.hyph` from https://github.com/koreader/koreader/issues/8843 (by making a `txt` file out of the `hyph` file and multiple copy and paste some of the words in that txt-file,; see attachment
[greek-text.txt](https://github.com/koreader/koreader/files/8191491/greek-text.txt)
)

> Would it make sense to add that upper dot (·) as a punctuation mark like (, ; ...) in crengine?

As a funny fact, with that crengine PR the upper dot `·` won't be part of a word anymore :)

**Comment by zwim** [comment]
Double posting as in https://github.com/koreader/crengine/pull/466
(I would tend to use NFKC everywhere (and drop NFC) as  utf8 casefold uses that too.)

**Comment by poire-z** [comment]
The french Wikipedia page gives some samples: https://fr.wikipedia.org/wiki/Normalisation_Unicode

> As a funny fact, with that crengine PR (https://github.com/koreader/crengine/pull/466) the upper dot · won't be part of a word anymore :)

Why is this change of behaviour happening ?

**Comment by poire-z** [comment]
Another thought: in most real books, I guess the text is most often normalized (I hope word editors end up normalizing what is inputed with multiple keystrokes).
Might this original issue be caused by the fact that to enter a composed letter with diacritics with our keyboard, we need to hit multiple keys, and the user hyph dict entry is so saved decomposed/non-normalized ? Or do this user hyph code normalize what's inputed ?

**Comment by zwim** [comment]
Until this PR we (almost) did not mess with the coding of a word. The only thing we have to do, is to lowercase the word. On saving an entry it is checked, if the lower cased entry matches the lower cased word. (So no need to think about our keyboard.)

And with that lower casing normalization come into the game. :(

**Comment by zwim** [comment]
> > As a funny fact, with that crengine PR ([koreader/crengine#466](https://github.com/koreader/crengine/pull/466)) the upper dot · won't be part of a word anymore :)
> 
> Why is this change of behaviour happening ?

I have absolutely no clue :(

**Comment by zwim** [comment]
Well, how shall we go on with this?

Currently I have disabled normalization, and added a NORM switch at the top of the file. (In the way crengine expects the dictionary).

Scrubbing itself does no harm and the sorting of entries is now as crengine would expect it.

When sometimes (if ever) normailzation is done with hyphenation in crengine, we just would have to set NORM as true and make sure that our `Utf8Proc.lowercase` uses the right normalization algorithm (NFKC, NFC) ...

**Comment by poire-z** [comment]
Looks ok to me.

File: frontend/apps/reader/modules/readeruserhyph.lua [hunk 8]
```diff
@@ -210,13 +304,13 @@ function ReaderUserHyph:modifyUserEntry(word)
...
                             -- don't save if no changes
-                            if Utf8Proc.lowercase(new_suggestion) ~= input_dialog.old_hyph_lowercase then
+                            if Utf8Proc.lowercase(new_suggestion, NORM) ~= input_dialog.old_hyph_lowercase then
                                 self:updateDictionary(word, new_suggestion)
                             end
                             UIManager:close(input_dialog)
                         else
                             UIManager:show(InfoMessage:new{
-                                text = T(_("Invalid hyphenation!"), self.dict_file),
+                                text = _("Invalid hyphenation!"),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readeruserhyph.lua:313]: Ignoring for a second that it wasn't used, that does look like useful info to stick on? But in that case more like:

```suggestion
                                text = _("Invalid hyphenation!") .. "\n" .. self.dict_file,
```

(Keep in mind I'm only skimming the diff though.)

**Comment by Frenzie** [comment]
@zwim Is this ready to merge?

File: frontend/apps/reader/modules/readeruserhyph.lua (hunk 8 above)
Comment (zwim) [reply frontend/apps/reader/modules/readeruserhyph.lua:313]: No, thats absolutely not necessary. Maybe something like
```
                           UIManager:show(InfoMessage:new{
                                text = T(_("Invalid hyphenation!\n%1\ndoes not match\n%2"),
                                    "    " .. word, "    " .. input_dialog:getInputText())
                            })
````
would give the user a more useful hint.

Comment (zwim) [reply frontend/apps/reader/modules/readeruserhyph.lua:313]: This will look like:
grafik

Comment (Frenzie) [reply frontend/apps/reader/modules/readeruserhyph.lua:313]: If the original input is normally visible as in your screenshot I don't think that's necessary.

Comment (zwim) [reply frontend/apps/reader/modules/readeruserhyph.lua:313]: Yes it is! -> So from my side it is ready to merge.

**Pull request merged**